### PR TITLE
Added Notify SDK

### PR DESCRIPTION
**Pull request by TalhaAli00**

### CODE REVIEW

**Review by Elyniss — APPROVED** [review]

File: core/android/src/main/kotlin/com/walletconnect/android/internal/common/di/DatabaseConfig.kt [hunk 1]
```diff
@@ -4,6 +4,8 @@ import com.walletconnect.utils.Empty
 import org.koin.android.ext.koin.androidContext
 import org.koin.core.scope.Scope
 
+// TODO: Convert this to an enum class so dbNames is built automatically
+@Suppress("PropertyName")
 class DatabaseConfig(private val storagePrefix: String = String.Empty) {
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply core/android/src/main/kotlin/com/walletconnect/android/internal/common/di/DatabaseConfig.kt:9]: ahh that's right. I'll remove the TODO

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/client/NotifyMessageService.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            }
+        )
+    }
+
+    override fun onMessageReceived(message: RemoteMessage) {
+        super.onMessageReceived(message)
+
+        with(message) {
+            try {
+                if (data.containsKey("topic") && data.containsKey("blob")) {
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/client/NotifyMessageService.kt:29]: added const vals

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/client/NotifyMessageService.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                        onError = { error -> onError(error.throwable, this) }
+                    )
+                } else if (notification?.isValid() == true) {
+                    val notifyMessage = with(requireNotNull(notification)) {
+                        Notify.Model.Message(
+                            title = title!!,
+                            body = body!!,
+                            icon = icon,
+                            url = imageUrl?.toString(),
+                            type = ""
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/client/NotifyMessageService.kt:43]: done

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/common/model/EngineDO.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import com.walletconnect.android.internal.common.model.AppMetaData
+import com.walletconnect.android.internal.common.model.Expiry
+import com.walletconnect.android.internal.common.model.RelayProtocolOptions
+import com.walletconnect.android.internal.common.model.type.EngineEvent
+import com.walletconnect.foundation.common.model.PublicKey
+import com.walletconnect.foundation.common.model.Topic
+
+internal sealed class EngineDO : EngineEvent {
+
+    sealed class Scope : EngineDO() {
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/common/model/EngineDO.kt:15]: Scope is used in several use-case files so I thought it made sense to keep the domain objects together. I'm open to ideas if we want to move this

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/di/NotifyEngineModule.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import com.walletconnect.notify.engine.NotifyEngine
+import com.walletconnect.notify.engine.domain.EngineNotifySubscriptionNotifier
+import com.walletconnect.notify.engine.domain.ExtractNotifyConfigUseCase
+import com.walletconnect.notify.engine.domain.RegisterIdentityAndReturnDidJwtUseCase
+import com.walletconnect.notify.engine.domain.RegisterIdentityAndReturnDidJwtUseCaseInterface
+import org.koin.core.qualifier.named
+import org.koin.dsl.module
+
+@JvmSynthetic
+internal fun walletEngineModule() = module {
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/di/NotifyEngineModule.kt:15]: done

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/sync/use_case/GetMessagesFromHistoryUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                val messagesBatchSize = 200L
+                historyInterface.getMessages(
+                    MessagesParams(topic.value, null, messagesBatchSize, null),
+                    onError = { error -> onError(error.throwable) },
+                    onSuccess = {
+                        //todo: Support fetching more than [messagesBatchSize]
+                        if (it.size >= messagesBatchSize) logger.error("Fetched $messagesBatchSize for ${store.value}") else logger.log("Fetched ${it.size} for ${store.value}")
+                        countDownLatch.countDown()
+                    }
+                )
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/sync/use_case/GetMessagesFromHistoryUseCase.kt:46]: sounds good

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/domain/ExtractNotifyConfigUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        val notifyConfigDappUri = dappUri.run {
+            if (this.path?.contains(WC_NOTIFY_CONFIG_JSON) == false) {
+                this.buildUpon().appendPath(WC_NOTIFY_CONFIG_JSON)
+            } else {
+                this
+            }
+        }
+
+        val wellKnownNotifyConfigString = URL(notifyConfigDappUri.toString()).openStream().bufferedReader().use { it.readText() }
+        val notifyConfig = serializer.tryDeserialize<NotifyConfigDTO>(wellKnownNotifyConfigString) ?: return@withContext Result.failure(Exception("Failed to parse well-known/wc-notify-config.json"))
```
Comment (kacperoak) [inline protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/domain/ExtractNotifyConfigUseCase.kt:25]: Maybe `"Failed to parse $WC_NOTIFY_CONFIG_JSON"`?

**Review by kacperoak — APPROVED** [review]

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/domain/ExtractNotifyConfigUseCase.kt (hunk 7 above)
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/domain/ExtractNotifyConfigUseCase.kt]: done

File: protocol/notify/src/main/kotlin/com/walletconnect/notify/common/model/EngineDO.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/notify/src/main/kotlin/com/walletconnect/notify/common/model/EngineDO.kt]: moved models out

**Pull request merged**